### ISSUE TITLE
Move regex-based History exclusion to AST-based to play nice with SecretMangement

### ISSUE THREAD

**Issue by TylerLeonhardt** [PowerShell/PSReadLine#1840]
## Description of the new feature/enhancement

Because every command in `SecretManagement` has `Secret` in the name, every command invocation gets excluded from history. This isn't a good experience especially since SecretManagement is helping with the "secrets not in plaintext" situation.

Here are some examples that would be nice to have in history:

```pwsh
Get-Secret PSGalleryApiKey -AsPlainText
```

```pwsh
Publish-Module -Name Foo -NuGetApiKey (Get-Secret PSGalleryApiKey -AsPlainText)
```

However, we need to make sure that something like this still gets excluded:

```pwsh
Get-ResultFromTwoSecrets -Secret1 (Get-Secret -Name blah -AsPlainText) -Secret2 sdv87ysdfayf798hfasd8f7ha
```

## Proposed technical implementation details (optional)

This regex based filter:
https://github.com/PowerShell/PSReadLine/blob/fd9980642ce7943510dc65ea974d45faa4c900af/PSReadLine/History.cs#L115

should be AST based and play nice with an exclusion list.

**Comment by daxian-dbw** [comment]
Also got feedback from @LeeHolmes about the regex used to filter out sensitive command lines:
> It looks like this is the list that SpecterOps uses to dig through command line history for sensitive content:
>
>[Seatbelt/PowerShellHistory.cs at master · GhostPack/Seatbelt · GitHub](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGhostPack%2FSeatbelt%2Fblob%2Fmaster%2FSeatbelt%2FCommands%2FWindows%2FPowerShellHistory.cs%23L17&data=04%7C01%7Cdongbow%40microsoft.com%7Cc17513e973e54b9d1ab408d8b74d39e6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637460889632388166%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=MHM8r7XQXzNlcpz3RoSAsFrbhU2ovJCwNTox37VD3uE%3D&reserved=0)
>[Seatbelt/MiscUtil.cs at cc3bde6a528cffe9dbc613906cc6d9b68278367a · GhostPack/Seatbelt · GitHub](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGhostPack%2FSeatbelt%2Fblob%2Fcc3bde6a528cffe9dbc613906cc6d9b68278367a%2FSeatbelt%2FUtil%2FMiscUtil.cs%23L89&data=04%7C01%7Cdongbow%40microsoft.com%7Cc17513e973e54b9d1ab408d8b74d39e6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637460889632388166%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=0R0%2BLO1h6vLTamlCMKM0v3K9lsrRmhLc%2B0Ngvd3JC3M%3D&reserved=0)
>
>I didn’t realize how many tools supported “/p <password>” and “-p <password>” and variants.
>
>Should we update this? PSReadLine/History.cs at fd1265e385cb4b7b860d8f1452da1ceb347e123e · PowerShell/PSReadLine · GitHub

**Comment by iSazonov** [comment]
Maybe add an attribute to sensitive parameters? This could helps to filter. Also we could not filter whole command string but only sensitive _argument_.

**Comment by bwright86** [comment]
I've been using Lee's suggested regex for a while, and I noticed it is a bit aggressive, not just with "Secret", but even if I pass a [PSCredential] object, it still gets yanked.

I think the scenario for the regex pattern is for parameters that accept a plaintext string of tokens/passwords/secret.

I think AST parsing is probably the best way to go, so you know if a string is passed.

But for now, i have modified the regex a little bit to explicitly look for a parameter and the start of a string:
$sensitive = "\b-.*(password|asplaintext|token|key|secret).*\b\s+[`"|']"

This appears to be less aggressive, and still removes plaintext strings, but allows for variables to be used and history kept.

**Comment by ghost** [comment]
:tada: This issue was addressed in 2921, which has now been successfully released in [`v2.2.0-beta4`](https://github.com/PowerShell/PSReadLine/releases/tag/v2.2.0-beta4). :tada:
